File: pytgm/geky.py
def geky(times=1):
    try:
        from msvcrt import getch as g
        for i in range(times):
            k = g()
            if k == b'\xe0':
                k = g()
                if k == b'H':
                    return 'ArrowUp'
                elif k == b'P':
                    return 'ArrowDown'
                elif k == b'K':
                    return 'ArrowLeft'
                elif k == b'M':
                    return 'ArrowRight'
            else:
                return key.decode()

    except:
        from sys import stdin
        from tty import setraw
        from termios import tcsetattr, TCSADRAIN, tcgetattr
        for i in range(times):
            fd = stdin.fileno()
            old = tcgetattr(fd)
            try:
                setraw(fd)
                c = stdin.read(1)
                if c == '\x1b':
                    c = stdin.read(2)
                    if c == '[A': return 'ArrowUp'
                    elif c == '[B': return 'ArrowDown'
                    elif c == '[C': return 'ArrowRight'
                    elif c == '[D': return 'ArrowLeft'
                    else: return c
                else:
                    if c == ' ': return 'space'
                    else: return c
            finally: tcsetattr(fd, TCSADRAIN, old)

File: pyTGM/terminal/geky/geky.cpp
#include "geky.hpp"

#include <iostream>
#include <string>
#include <stdexcept>

#ifdef _WIN32
#include <windows.h>
#include <conio.h>
#else
#include <termios.h>
#include <unistd.h>
#include <fcntl.h>
#endif

std::string geky(int times = 1) {
    std::string result;

#ifdef _WIN32
    // Windows-specific key reading
    for (int i = 0; i < times; ++i) {
        char c = _getch();  // Read a single character
        if (c == 0 || c == -32) {  // Handle special keys
            c = _getch();  // Read the special key code
            switch (c) {
                case 72: result = "ArrowUp"; break;
                case 80: result = "ArrowDown"; break;
                case 75: result = "ArrowLeft"; break;
                case 77: result = "ArrowRight"; break;
            }
        } else {
            result = c;  // Regular character
        }
    }
#else
    // Unix-like (Linux/macOS) key reading
    for (int i = 0; i < times; ++i) {
        struct termios oldt, newt;
        tcgetattr(STDIN_FILENO, &oldt);
        newt = oldt;
        newt.c_lflag &= ~(ICANON | ECHO);  // Disable canonical mode and echo
        tcsetattr(STDIN_FILENO, TCSANOW, &newt);
        char c;
        read(STDIN_FILENO, &c, 1);
        tcsetattr(STDIN_FILENO, TCSANOW, &oldt);

        if (c == 27) {  // Handle escape sequence for arrow keys
            char seq[2];
            read(STDIN_FILENO, seq, 2);
            if (seq[0] == '[') {
                switch (seq[1]) {
                    case 'A': result = "ArrowUp"; break;
                    case 'B': result = "ArrowDown"; break;
                    case 'C': result = "ArrowRight"; break;
                    case 'D': result = "ArrowLeft"; break;
                }
            }
        } else {
            result += c;  // Regular character
        }
    }
#endif

    return result;
}
